ports/stm32: Reserve SPI flash bus.

iabdalkader opened the pull request:
### Summary

Reserved SPI buses must remain initialized during a soft reboot as they may be used for SPI flash storage or XIP. Fixes an issue introduced in #16346.

iabdalkader committed:
stm32/spi: Retain the state of special SPI buses on soft reboot.

Reserved and static SPI buses must remain initialized during a soft reboot
as they may be used for SPI flash storage or XIP.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32/boards: Reserve SPI bus when it's used for external flash storage.

Reserve SPI flash bus used for storage or XIP.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

dpgeorge commented:
I've tested this on LEGO_HUB_NO6, and it fixes the problem (soft reset hangs).

But I don't think the approach will work.  The `MIKROE_QUAIL` board uses SPI3 for two of its click connectors as well as the external SPI flash.  So this PR will be a big breaking change for that board, users will no longer be able to use SPI on two of the click connectors.

Instead I suggest a `MICROPY_HW_SPI1_IS_STATIC(spi_id)` macro which a board can use to say that certain SPI buses should not be deinitialized.

iabdalkader commented:
> The `MIKROE_QUAIL` board uses SPI3 for two of its click connectors as well as the external SPI flash.
>  So this PR will be a big breaking change for that board, users will no longer be able to use SPI on two of the click connectors.

I could just add the static macro and use it for this particular board, but I don't think this SPI should be shared. If allow it to be used from machine_spi users could reconfigure it changing the polarity/phase or baud would would break the SPI flash.

iabdalkader committed:
stm32/mpconfigboard_common: Add MICROPY_HW_SPI_IS_STATIC macro.

A board should make this return true if the specified SPI instances should
not be deinitialized on soft-reboot.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

dpgeorge reviewed (approved):
Looks good now, thank you.

I tested again on LEGO_HUB_NO6 and it still works.
